File: python/src/pipelines/zebra_wings/embeddings/presaved_embeddings_initializer.py
import keras
import numpy as np

class PreSavedEmbeddingsInitializer(keras.initializers.Initializer):
    def __init__(self, pretrained_ngram_vectors=None):
        self.config = { "pretrained_ngram_vectors": pretrained_ngram_vectors }

    def __call__(self, shape, dtype=None):
        if not self.config["pretrained_ngram_vectors"]:
            return keras.constant(0, shape=shape, dtype=dtype)
        else:
            #  pretrained_ngram_vectors is like [['__', [0, ... ]]], so we only extract the vectors with index order
            encoded_sentences = [x[1] for x in self.config["pretrained_ngram_vectors"]]
            return np.array(encoded_sentences, dtype)

    def get_config(self):
        return self.config["pretrained_ngram_vectors"]

File: typescript/src/pipelines/zebraWings/embeddings/PreSavedEmbeddingsInitializer.ts
import * as tf from '@tensorflow/tfjs';
import * as initializers from '@tensorflow/tfjs-layers/dist/initializers';
import { flatMapDeep } from 'lodash';

export interface IEmbeddingsModelConfig {
    pretrainedNGramVectors: Map<string, Float32Array>;
    embeddingDimensions: number;
}

export class PreSavedEmbeddingsInitializer extends initializers.Initializer {
    public static className = 'PreSavedEmbeddingsInitializer';
    public config: IEmbeddingsModelConfig;
    public className = PreSavedEmbeddingsInitializer.className;
    constructor(config: IEmbeddingsModelConfig) {
        super();
        this.config = config;
    }
    public apply(shape: tf.Shape, dtype: tf.DataType): tf.Tensor {
        if (!this.config || !this.config.pretrainedNGramVectors) {
            return tf.zeros(shape, dtype);
        }
        return tf.tidy(() => {
            const flatMat = flatMapDeep([...this.config.pretrainedNGramVectors.values()]);
            return tf.tensor2d(flatMat, [this.config.pretrainedNGramVectors.size, this.config.embeddingDimensions], 'float32');
        });
    }

    public getConfig() {
        return this.config.pretrainedNGramVectors as any;
    }
}
tf.serialization.SerializationMap.register(PreSavedEmbeddingsInitializer);
